Simplify BenchmarkBatchStreamReaders by introducing Params

yingsu00 opened the pull request:
```
== NO RELEASE NOTE ==
```

yingsu00 commented:
@mbasmanova if you like the idea using Nulls enum and switch statements on typeSignature to additionally simplify BenchmarkSelectiveStreamReaders I can apply the same change to it.

mbasmanova commented on presto-orc/src/test/java/com/facebook/presto/orc/BenchmarkBatchStreamReaders.java:
```diff
@@ -87,683 +82,118 @@
-                else {
-                    values.add(null);
-                }
+            if (withNulls == Nulls.ALL) {
```
nit: switch statement would be more natural here

```
            switch (withNulls) {
                case ALL:
                    return NULL_VALUES;
                case PARTIAL:
                    return IntStream.range(0, ROWS).mapToObj(i -> i % 2 == 0 ? createValue() : null).collect(toList());
                default:
                    return IntStream.range(0, ROWS).mapToObj(i -> createValue()).collect(toList());
            }
```

yingsu00 commented:
@mbasmanova 

> > Add benchmark for SliceBatchDirectStreamReader
> 
> The commit message seems to contradict the changes that appear to add benchmarks for both direct and dictionary readers.

The existing "varchar" typeSignature only tested SliceDictionaryBatchStreamReader. This commit renamed it to "varchar_dictionary" and added a new "varchar_direct" typeSignature to test SliceBatchDirectStreamReader by creating random String values. How about adding explanation to the commit message?

```
Add benchmark for SliceBatchDirectStreamReader

The existing "varchar" typeSignature benchmark only tested
SliceDictionaryBatchStreamReader. This commit renamed it to
"varchar_dictionary" and added a new "varchar_direct" typeSignature to
test SliceBatchDirectStreamReader.
```

yingsu00 commented:
> > Use per-row metric in BenchmarkBatchStreamReaders
> 
> It is hard to understand the effect of this change. Can you show the results before and after the change and explain the differences?

The benefit of using NANOSECOND as output unit and amortizing results by number of rows is

1. Better precision: it allows for more digits to show in the result
2. Per-row output is amortized by ROWS and this makes the result independent of total rows and easier to compare the results across benchmarks.

An example before the change:

```

Iteration   1: 0.036 s/op
Iteration   2: 0.038 s/op
...
Iteration  20: 0.035 s/op

Result "com.facebook.presto.orc.BenchmarkBatchStreamReaders.readBlocks":
  0.039 ±(99.9%) 0.001 s/op [Average]
  (min, avg, max) = (0.035, 0.039, 0.046), stdev = 0.003
  CI (99.9%): [0.038, 0.041] (assumes normal distribution)


# Run complete. Total time: 00:01:14

Benchmark                               (typeSignature)  (withNulls)  Mode  Cnt  Score   Error  Units
BenchmarkBatchStreamReaders.readBlocks          boolean      PARTIAL  avgt   60  0.039 ± 0.001   s/op

```

Example After change:

```
Iteration   1: 3.694 ns/op
Iteration   2: 3.556 ns/op
...
Iteration  20: 3.515 ns/op

Result "com.facebook.presto.orc.BenchmarkBatchStreamReaders.readBlocks":
  3.940 ±(99.9%) 0.192 ns/op [Average]
  (min, avg, max) = (3.515, 3.940, 5.959), stdev = 0.429
  CI (99.9%): [3.749, 4.132] (assumes normal distribution)


# Run complete. Total time: 00:01:16

Benchmark                               (typeSignature)  (withNulls)  Mode  Cnt  Score   Error  Units
BenchmarkBatchStreamReaders.readBlocks          boolean      PARTIAL  avgt   60  3.940 ± 0.192  ns/op
```

How does the the following commit message look?

```
Use per-row metric in BenchmarkBatchStreamReaders

Using NANOSECOND as output unit and amortizing results by number of rows
allows for better precision and easier results comparison across benchmarks.

An example output before the change:
Benchmark                               (typeSignature)  (withNulls)  Mode  Cnt  Score   Error  Units
BenchmarkBatchStreamReaders.readBlocks          boolean      PARTIAL  avgt   60  0.039 ± 0.001   s/op

An example output after the change:
Benchmark                               (typeSignature)  (withNulls)  Mode  Cnt  Score   Error  Units
BenchmarkBatchStreamReaders.readBlocks          boolean      PARTIAL  avgt   60  3.940 ± 0.192  ns/op

```

yingsu00 commented:
> > Avoid creating OrcDataSource and OrcReader in each iteration
> 
> Why is this useful?

Before this commit, createRecordReader() creates new OrcDataSource and OrcReader in every iteration. However this cost is not what we are interested and should be minimized. Making them class members of BenchmarkData would make such object creation happen only once per fork when the setup() method is called, and makes the result slightly more accurate.  

Example result before this change:
```
Benchmark                               (typeSignature)  (withNulls)  Mode  Cnt  Score   Error  Units
BenchmarkBatchStreamReaders.readBlocks          boolean      PARTIAL  avgt   60  3.940 ± 0.192  ns/op

```
Example after this change shows slightly better result
```
Benchmark                               (typeSignature)  (withNulls)  Mode  Cnt  Score   Error  Units
BenchmarkBatchStreamReaders.readBlocks          boolean      PARTIAL  avgt   60  3.814 ± 0.077  ns/op
```

I also updated the commit message as follows:
```
Avoid creating OrcDataSource and OrcReader in each iteration

Minimize the cost of createRecordReader() in each iteration by making
OrcDataSource and OrcReader class member of BenchmarkData makes the
reuslt more accurate.
```
Please let me know if it needs more polish-up.

mbasmanova commented:
@yingsu00 Thanks for detailed explanations. The first 4 commits look good. These are helpful refactoring that don't introduce material changes to the benchmark. The last 2 two commit do modify the benchmark itself. My understanding is that now the benchmark is measuring cost of reading one row as opposed to cost of reading a whole "split". This could be useful, but I don't want to loose the original benchmark. First, with these changes we can no longer use the benchmark to compare batch and selective stream readers. Second, we no longer take into account "setup costs", which are important. Is there any particular reason for introducing changes in the last 2 commits?

yingsu00 commented:
> @yingsu00 Thanks for detailed explanations. The first 4 commits look good. These are helpful refactoring that don't introduce material changes to the benchmark. The last 2 two commit do modify the benchmark itself. My understanding is that now the benchmark is measuring cost of reading one row as opposed to cost of reading a whole "split". This could be useful, but I don't want to loose the original benchmark. First, with these changes we can no longer use the benchmark to compare batch and selective stream readers. Second, we no longer take into account "setup costs", which are important. Is there any particular reason for introducing changes in the last 2 commits?

@mbasmanova  Hi Masha, thanks for the explanation! For your question of the 5th commit "Use per-row metric in BenchmarkBatchStreamReaders" we can apply to BenchmarkSelectiveStreamReaders too. Then the comparison doesn't reply on ROWS and is more generic.

For the last commit, I didn't realize creating OrcReader is also part of the target of measurement. But is creating FileOrcDataSource also part of the measurement target? If not we should avoid creating it every time. If yes then we should keep it.

yingsu00 commented:
The travis run has already succeeded.
